src/KafkaFlow/Event.cs

```diff
@@ -0,0 +1,63 @@
+﻿namespace KafkaFlow
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    internal class Event<TArg> : IEvent<TArg>
+    {
+        private readonly ILogHandler logHandler;
+
+        private readonly IList<Func<TArg, Task>> handlers = new List<Func<TArg, Task>>();
+
+        public Event(ILogHandler logHandler)
+        {
+            this.logHandler = logHandler;
+        }
+
+        public void Subscribe(Func<TArg, Task> handler)
+        {
+            this.handlers.Add(handler);
+        }
+
+        internal async Task FireAsync(TArg arg)
+        {
+            foreach (var handler in this.handlers)
+            {
+                try
+                {
+                    if (handler is null)
+                    {
+                        continue;
+                    }
+
+                    await handler.Invoke(arg);
+                }
+                catch (Exception e)
+                {
+                    this.logHandler.Error("Error firing event", e, new { Event = this.GetType().Name });
+                }
+            }
+        }
+    }
+
+    internal class Event : IEvent
+    {
+        private readonly Event<object> evt;
+
+        public Event(ILogHandler logHandler)
+        {
+            this.evt = new Event<object>(logHandler);
+        }
+
+        public void Subscribe(Func<Task> handle)
+        {
+            this.evt.Subscribe(_ => handle?.Invoke());
+        }
+
+        internal Task FireAsync()
```

Comment:
Should we also create a Fire() method for non async callers?

Reply:
The event is fired from inside of the framework and we don't know if the handler will be sync or async, it will depends on the specific client logic.
For that, it is better to work always with async and the handler can implement any logic there. (If it is not async they can returns Task.CompletedTask)